pkg/llbutil/id.go

```diff
@@ -34,13 +40,60 @@ func LocalID(ctx context.Context, absPath string, opts ...llb.LocalOption) (stri
 	return digest.FromBytes(def.Def[len(def.Def)-1]).String(), nil
 }
 
-// LocalUniqueID returns a consistent string that is unique per host + cwd
-func LocalUniqueID(cwd string) (string, error) {
+// localUniqueID returns a consistent string that is unique per host + dir +
+// last modified time.
+//
+// If there is already a solve in progress using the same local dir, we want to
+// deduplicate the "local" if the directory hasn't changed, but if there has
+// been a change, we must not identify the "local" as a duplicate. Thus, we
+// incorporate the last modified timestamp into the result.
+func localUniqueID(localPath string, opts ...llb.LocalOption) (string, error) {
 	mac, err := FirstUpInterface()
 	if err != nil {
 		return "", err
 	}
-	return fmt.Sprintf("cwd:%s,mac:%s", cwd, mac), nil
+
+	var localInfo llb.LocalInfo
+	for _, opt := range opts {
+		opt.SetLocalOption(&localInfo)
+	}
+
+	var walkOpts fsutil.WalkOpt
+	if localInfo.IncludePatterns != "" {
+		if err := json.Unmarshal([]byte(localInfo.IncludePatterns), &walkOpts.IncludePatterns); err != nil {
+			return "", errors.Wrap(err, "failed to unmarshal IncludePatterns for localUniqueID")
+		}
+	}
+	if localInfo.ExcludePatterns != "" {
+		if err := json.Unmarshal([]byte(localInfo.ExcludePatterns), &walkOpts.ExcludePatterns); err != nil {
+			return "", errors.Wrap(err, "failed to unmarshal ExcludePatterns for localUniqueID")
+		}
+	}
+	if localInfo.FollowPaths != "" {
+		if err := json.Unmarshal([]byte(localInfo.FollowPaths), &walkOpts.FollowPaths); err != nil {
+			return "", errors.Wrap(err, "failed to unmarshal FollowPaths for localUniqueID")
+		}
+	}
+
+	fi, err := os.Stat(localPath)
+	if err != nil {
+		return "", err
+	}
+
+	lastModified := fi.ModTime()
+	if fi.IsDir() {
```

Comment:
nit, we dont need to inspect the opts and/or unmarshal teh localnfo configs in the !IsDir case